libgadget/fof.c

```diff
@@ -442,33 +424,38 @@ fofp_merge(int target, int other, TreeWalk * tw)
 {
     /* this will lock h1 */
     int * Head = FOF_PRIMARY_GET_PRIV(tw)->Head;
-    struct SpinLocks * spin = FOF_PRIMARY_GET_PRIV(tw)->spin;
     int h1, h2;
-
     do {
-        h1 = HEADl(-1, target, -1, Head, spin);
-        /* stop looking if we find h1 along the path (because it is already owned by us) */
-        h2 = HEADl(h1, other, h1, Head, spin);
-        /* We had a lock already taken on h2 by another thread.
-         * We need to unlock h1 and retry to avoid deadlock loops.*/
-        if(h2 == -2)
-            unlock_spinlock(h1, spin);
-    } while(h2 == -2);
-
-    if(h2 >=0)
-    {
-        /* h2 as a sub-tree of h1 */
-        Head[h2] = h1;
-
-        /* update MinID of h1 */
-        if(HaloLabel[h1].MinID > HaloLabel[h2].MinID)
-        {
-            HaloLabel[h1].MinID = HaloLabel[h2].MinID;
-            HaloLabel[h1].MinIDTask = HaloLabel[h2].MinIDTask;
+        h1 = HEADl(-1, target, Head);
+        /* Done if we find h1 along the path
+         * (because other is already in the same halo) */
+        h2 = HEADl(h1, other, Head);
+        if(h2 < 0)
+            return;
+        /* Ensure that we always merge to the lower entry*/
+        if(h1 > h2) {
+            int tmp = h2;
+            h2 = h1;
+            h1 = tmp;
         }
-        //printf("unlocking %d by %d in merge\n", h2, omp_get_thread_num());
-        unlock_spinlock(h2, spin);
+     /* Atomic compare exchange to make h2 a subtree of h1.
+      * Set Head[h2] = h1 iff Head[h2] is still h2. Otherwise loop.*/
+    } while(!__atomic_compare_exchange(&Head[h2], &h2, &h1, 0, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
```

Comment:
Do we need to mark Head, h2 and/or h1 volatile?

Reply:
I don't think so - the C standard doesn't seem to require it with atomic_compare_exchange. It would be Head that would be volatile, as the others are local variables. I understood that volatile basically means that the memory may be altered by another process and is used to avoid the compiler optimizing out pointless loops. I think that with openmp threads it is not necessary.